### PR TITLE
added a test to make sure all files are using snake_case

### PR DESCRIPTION
**Pull request by smeevil**
https://github.com/lpil/dogma/issues/185

### CODE REVIEW

File: lib/dogma/rule/snake_case_filename.ex [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    desired_filename = filename
+      |> String.replace(~r/([A-Z])/, "_\\g{1}")
+      |> String.replace(~r/\W/, "_")
+      |> String.replace(~r/\A_/, "")
+      |> String.replace(~r/[_]+/, "_")
+      |> String.downcase
+
+    desired_path = reconstruct_path(dirname, desired_filename, ext)
+
+    if String.equivalent?(path, desired_path) do
```
Comment (lpil) [inline lib/dogma/rule/snake_case_filename.ex]: Where does this function come from? Seems undefined

File: lib/dogma/rule/snake_case_filename.ex [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  aValidFilename.ex
+  a valid filename.ex
+  a.valid.filename.ex
+  a-valid-filename.ex
+  """
+
+  def test(_, %{path: path}) do
+    {dirname, filename, ext} = deconstruct_path(path)
+
+    desired_filename = filename
```
Comment (lpil) [inline lib/dogma/rule/snake_case_filename.ex]: There is already a util module for attempting to detect whether a name is snake_case, use that instead.

If you can demonstrate that this implementation is better you can replace the implementation with this one :)

**Comment by lpil** [comment]
Cool! A few comments above.

File: test/dogma/rule/snake_case_filename_test.exs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+defmodule Dogma.Rule.SnakeCaseFilenameTest do
+  use RuleCase, for: SnakeCaseFilename
+
+  should "not generate errors for snake cased files" do
+    result = Script.parse!("", "this_is_correct.ex")
+
```
Comment (lpil) [inline test/dogma/rule/snake_case_filename_test.exs]: Remove these newline please

File: test/dogma/rule/snake_case_filename_test.exs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    assert expected_error == Rule.test(@rule, result)
+  end
+
+  should "error when a file contains invalid characters" do
+    result = Script.parse!("", "ThisIs-so.Wrong.ex")
+
+    expected_error = [%Dogma.Error{
+      line: 0,
+      message: "Rename ThisIs-so.Wrong.ex to this_is_so_wrong.ex",
```
Comment (lpil) [inline test/dogma/rule/snake_case_filename_test.exs]: Should we test each edge case separately?

File: lib/dogma/rule/snake_case_filename.ex (hunk 1 above)
Comment (smeevil) [inline lib/dogma/rule/snake_case_filename.ex]: not sure which function you mean ?

File: lib/dogma/rule/snake_case_filename.ex (hunk 2 above)
Comment (smeevil) [inline lib/dogma/rule/snake_case_filename.ex]: Using the internal tool for detecting snake case now.

File: test/dogma/rule/snake_case_filename_test.exs (hunk 4 above)
Comment (smeevil) [inline test/dogma/rule/snake_case_filename_test.exs]: We could do that, but I don't really see the added value to that.
If you like give me a few cases you would like to test and I'll add them.
